source/common/common/envoy_defaults.h

```diff
@@ -0,0 +1,106 @@
+    if (msg.has_##field_name()) {                                                                  \
+      return msg.field_name().value();                                                             \
+    }                                                                                              \
+    return DefaultsProfile::get().getNumber(config, #field_name, default_value);                   \
+  }((message)))
+
+#define PROTOBUF_GET_MS_OR_PROFILE_DEFAULT(message, config, field_name, default_value)             \
+  ([&](const auto& msg) -> std::chrono::milliseconds {                                             \
+    if (msg.has_##field_name()) {                                                                  \
+      return std::chrono::milliseconds(DurationUtil::durationToMilliseconds(msg.field_name()));    \
+    }                                                                                              \
+    return DefaultsProfile::get().getMs(config, #field_name, default_value);                       \
+  }((message)))
+
+#define PROTOBUF_GET_BOOL_OR_PROFILE_DEFAULT(message, config, field_name, default_value)           \
+  ([&](const auto& msg) -> bool {                                                                  \
+    if (msg.has_##field_name()) {                                                                  \
+      return msg.field_name().value();                                                             \
+    }                                                                                              \
+    return DefaultsProfile::get().getBool(config, #field_name, default_value);                     \
+  }((message)))
+
+namespace Envoy {
+
+struct DefaultsProfile {
+  enum Profile {
+    Performant,
+    Safe,
+  };
+
+  class ConfigContext {
+  public:
+    ConfigContext() = default;
+    ConfigContext(const ConfigContext& context) = default;
+    ConfigContext(const Protobuf::Message& config) : ctx_(config.GetDescriptor()->full_name()) {}
+    // ConfigContext(const ConfigContext& context) : ctx_(context.ctx_) {}
+    ConfigContext(const std::string& context) : ctx_(context) {}
+    const absl::string_view getContext() const { return ctx_; }
+    ConfigContext& appendField(absl::string_view field) {
+      absl::StrAppend(&ctx_, ".", field);
+      return *this;
+    }
+
+  private:
+    std::string ctx_;
+  };
+
+  /**
+   * DefaultsProfile intended to be constructed as a DefaultsProfileSingleton, so that default
+   * values can be accessed globally.
+   * @param profile enum value specifying which profile to load
+   */
+  DefaultsProfile(Profile profile = Performant);
+  /**
+   * Returns a reference to DefaultsProfile inside singleton if it exists, otherwise returns
+   * blank DefaultsProfile.
+   */
+  static const DefaultsProfile& get();
+
+  /**
+   * Getters return the value associated with `field` in the DefaultsProfile if it exists,
```

Comment:
nit: the method comment should be repeated for each method, as it will be shown in IDEs.